File: lib/github_contributions.rb
#!/usr/bin/env ruby

module GithubContributions
  require 'octokit'

  REPOS = [
    'rails/rails', 'aws/chalice',
    'ajaxorg/ace', 'jneen/rouge', 'gnocchixyz/gnocchi',
    'gnocchixyz/python-gnocchiclient', 'ElemeFE/element', 'iview/iview-doc',
    'openstack-dev/pbr',
  ]

  class << self
    def login
      Octokit.configure do |c|
        c.login = ENV['GITHUB_LOGIN']
        c.password = ENV['GITHUB_PASSWORD']
      end
      puts "Logged in as #{Octokit.user.login}"
    end

    def my_contributions
      login
      REPOS.map { |repo_url| Repository.new(repo_url) }
    end

    def write_json(contributions)
      path = '/home/ubuntu/workspace/portfolio/data/open_source.json'
      File.open(path, 'w') do |f|
        f.puts contributions.map(&:to_json).to_json
      end
    end
  end

  class Repository
    attr_reader :url, :commits, :additions, :deletions

    def initialize(repository_url)
      puts "Obtaining contributions for #{repository_url}"

      @url = repository_url
      @description = Octokit.repository(@url).description
      @stats ||= retrieve_stats
      @commits = Octokit.commits(@url, author: ENV['GITHUB_LOGIN']).count

      unless @stats.nil?
        @additions = @stats.weeks.map { |week| week[:a] }.reduce(:+)
        @deletions = @stats.weeks.map { |week| week[:d] }.reduce(:+)
      end
    end

    def to_json
      {
        'url' => @url,
        'description' => @description,
        'commits' => @commits,
        'additions' => @additions,
        'deletions' => @deletions,
        'image' => @url.split('/').last
      }
    end

    private

    def retrieve_stats
      # If the data hasn't been cached when you query a repository's statistics,
      # you'll receive a 202 response; a background job is also fired to start
      # compiling these statistics. Give the job a few moments to complete, and
      # then submit the request again. If the job has completed, that request
      # will receive a 200 response with the statistics in the response body.
      begin
        stats = Octokit.contributors_stats(@url).find do |c|
          c.author.login == ENV['GITHUB_LOGIN']
        end
      rescue NoMethodError, 'No successful response. Re-trying in 10 seconds.'
        sleep(10)
        retry
      end

      stats
    end
  end
end

puts 'Obtaining contributions from Github...'
GithubContributions.write_json(GithubContributions.my_contributions)

puts 'Done'

File: lib/github_contributions.py
import json
import os

from dotenv import load_dotenv
from github import Github

load_dotenv()


USERNAME = 'BigChief45'
OUTPUT_PATH = os.path.join(os.path.dirname(__file__), '../data/open_source.json')

REPOSITORIES = [
  'rails/rails',
  'aws/chalice',
  'aws/aws-sdk-java',
  'serverless/serverless',
  'encode/httpx',
  'ajaxorg/ace',
  'timgrossmann/InstaPy',
  'jneen/rouge',
  'gnocchixyz/gnocchi',
  # 'gnocchixyz/python-gnocchiclient',
  # 'ElemeFE/element',
  # 'iview/iview-doc',
  'openstack-dev/pbr',
  # 'GetStream/Winds',
]


client = Github(os.environ['GITHUB_TOKEN'])


class Repository:

    def __init__(self, url):
        self._repository = client.get_repo(url)
        self.url = url
        self.description = self._repository.description

        self._stats = self._repository.get_stats_contributors()

    def _get_stats_for_user(self, username: str):
        # Stats for a repository will be empty if not enough contributions
        # have been made.

        user_stats = None
        for stat in self._stats:
            if stat.author.login == username:
                user_stats = stat
                break

        return user_stats

    def get_user_commits(self, username: str) -> int:
        return self._repository.get_commits(author=username).totalCount

    def get_user_additions(self, username: str) -> int:
        user_stats = self._get_stats_for_user(username)
        if user_stats is None:
            return None

        return sum([week.a for week in user_stats.weeks])

    def get_user_deletions(self, username: str) -> int:
        user_stats = self._get_stats_for_user(username)
        if user_stats is None:
            return None

        return sum([week.d for week in user_stats.weeks])

    def get_user_contributions(self, username: str):
        print(f'Getting contributions for repository {self.url}')

        return {
            'url': self.url,
            'description': self._repository.description,
            'commits': self.get_user_commits(username),
            'additions': self.get_user_additions(username),
            'deletions': self.get_user_deletions(username),
            'image': self.url.split('/')[-1]
        }


if __name__ == '__main__':
    contributions = [
        Repository(repo_url).get_user_contributions(USERNAME) for repo_url in REPOSITORIES
    ]

    with open(OUTPUT_PATH, 'w') as output_file:
        print(f'Saving contributions data to {OUTPUT_PATH}')
        output_file.write(json.dumps(contributions))
